### PR TITLE
Add restore function

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Subject

See https://github.com/sonata-project/SonataAdminBundle/pull/5807#issuecomment-573427323 for more information.
I am targeting this branch, because BC-break free.

## Changelog

```markdown
### Added
- Restore function in route collection
```

## To do
- [x] Update the tests
- [x] Update the documentation

### CODE REVIEW

File: src/Route/RouteCollection.php [hunk 1]
```diff
@@ -27,6 +27,11 @@ class RouteCollection
      */
     protected $elements = [];
 
+    /**
+     * @var Route[]
+     */
+    protected $cachedElements = [];
```
Comment (greg0ire) [inline src/Route/RouteCollection.php]:
```suggestion
    private $cachedElements = [];
```

File: src/Route/RouteCollection.php [hunk 2]
```diff
@@ -157,6 +162,16 @@ public function has($name)
         return \array_key_exists($this->getCode($name), $this->elements);
     }
 
+    /**
+     * @param string $name
+     *
+     * @return bool
+     */
+    public function hasCached($name)
```
Comment (greg0ire) [inline src/Route/RouteCollection.php]: You can and should use php 7 type declarations
```suggestion
    public function hasCached(string $name): bool
```

Comment (VincentLanglet) [reply src/Route/RouteCollection.php]: Isn't it weird having some public function final with php 7 type declaration and some without ? How to decide ?

Comment (greg0ire) [reply src/Route/RouteCollection.php]: > How to decide ?

You can use `final` and type declarations every time that's not a BC-break. So for all new methods, for instance.

Comment (VincentLanglet) [reply src/Route/RouteCollection.php]: Ok, I did it for every new function.

**Comment by greg0ire** [comment]
Thanks @VincentLanglet !

**Comment by VincentLanglet** [comment]
You're welcome @greg0ire. I would love a new release when you'll have time :)

**Pull request merged**